### PR TITLE
cleaned up the song json passed both to app and admin

### PR DESCRIPTION
**Pull request by ReganRyanNZ**
Big PR:

- A couple of refactors n comments, so the code is cleaner and easier to understand.
- Removed `model` from the song entries. All necessary attributes are on a flat array, and unused attributes have been cut from app and admin song lists.
- Separate lists for admin and app, since they need different attributes.
- Turned duplicates algorithm into a scope, and added "within 3 months" so old and consciously ignored duplicates wont show forever.
- Only super admins can see duplicates (since only they can remove songs anyway).

## UI
image

### CODE REVIEW

File: app/controllers/songs_controller.rb [hunk 1]
```diff
@@ -65,48 +66,52 @@ def adjust_lang_params
...
     @songs_to_check ||= {}
-    @songs_to_check[:changed] = Song.recently_changed.map { |s|
-      {
-        title: [s.custom_title, s.firstline_title, s.chorus_title].reject(&:blank?).first,
-        model: s,
-        edit_timestamp: s.updated_at
-      }
-    }
+    @songs_to_check[:changed] = sort_songs(Song.recently_changed.map { |song| admin_song_entry(song.titles.values.first, song) })
+    @songs_to_check[:duplicates] = sort_songs(Song.duplicates.map { |song| admin_song_entry(song.titles.values.first, song) }) if super_admin
```
Comment (MC-Squared) [inline app/controllers/songs_controller.rb]: `song.titles.values.first` could be replaced with `song.firstline_title` ? makes it a bit clearer and shorter.

Also, I think currently `@songs_to_check` are not removed from `@songs` - so any songs up the top will also appear further down in the list - not a big deal though.

File: app/controllers/songs_controller.rb [hunk 2]
```diff
@@ -65,48 +66,52 @@ def adjust_lang_params
...
+      books: song.book_indices,
+      edit_timestamp: song.updated_at,
+      last_editor: song.last_editor
+    }
+  end
+
+  def song_entry(title, song)
+    {
+      id: song.id,
+      title: title,
```
Comment (MC-Squared) [inline app/controllers/songs_controller.rb]: Do you think long-term it would be good to put chorus_title in here rather than duplicating the song twice? Makes the sorting a bit harder I guess. Maybe some other field like `redirect_to: song.id` so choruses can just point to the original song?
Anyway, not important for this PR, just thinking...

File: app/models/song.rb [hunk 3]
```diff
@@ -6,6 +6,14 @@ class Song < ApplicationRecord
 
   scope :audited, -> { joins(:audits).order('audits.time ASC') }
   scope :recently_changed, -> { where('updated_at >= ?', 1.week.ago).order(updated_at: :desc) }
+  scope :duplicates, -> {
+    where(
+      firstline_title: Song.select(:firstline_title)
+        .group(:firstline_title) # group songs into buckets of the same title
+        .having("count(*) > 1") # bucket has more than one song
+        .select(:firstline_title) # get firstline from these buckets
+    ).where("updated_at > ?", 3.months.ago) # old duplicates can be considered checked and ignored
```
[the remark this answers is not in the document]
Comment (ReganRyanNZ) [reply app/models/song.rb]: Some (dumb) songs have like entire first stanzas the same (Faith and Love album have songs like this).

Also, might want eventually to make duplicates fuzzy instead of exact, so that if someone types up the same song but the title has e.g. one character or capitalisation difference, it'll still show. If we get fuzzy, then there might be false positives in the duplicates (which will go away if ignored).

TL;DR—if for any reason there is a duplicate we want to keep, it's nice to have it not red eventually.

File: app/views/songs/_admin_songs.html.erb [hunk 4]
```diff
@@ -1,12 +1,12 @@
 <% songs.each do |song| %>
   <tr>
-    <td><%= link_to song[:title], edit_song_path(song[:model]), class: "wash-link edit_song_link", id: (review_changed ? "req_review_changed" : review_duplicate ? "req_review_duplicate" : nil) %></td>
+    <td><%= link_to song[:title], edit_song_path(song[:id]), class: "wash-link edit_song_link", id: (review_changed ? "req_review_changed" : review_duplicate ? "req_review_duplicate" : nil) %></td>
 
     <% if super_admin %>
-      <td><%= link_to 'Remove', song[:model], method: :delete, data: { confirm: 'Are you sure?' }, class: "wash-link remove_song_link" %></td>
+      <td><%= link_to 'Remove', song_path(song[:id]), method: :delete, data: { confirm: 'Are you sure?' }, class: "wash-link remove_song_link" %></td>
```
[the remark this answers is not in the document]
Comment (ReganRyanNZ) [reply app/views/songs/_admin_songs.html.erb]: really interestingly that works (even though `song` here isn't actually an activerecord), but it also throws the rest of its fields as GET params, so its nicer to leave with the id

File: app/controllers/songs_controller.rb (hunk 1 above)
Comment (ReganRyanNZ) [reply app/controllers/songs_controller.rb]: `song.titles.values.first` lets custom titles take precedence

you're right about the overlap in lists, I'll add that to the trello board

File: app/controllers/songs_controller.rb (hunk 2 above)
Comment (ReganRyanNZ) [reply app/controllers/songs_controller.rb]: yea, multiplying records for each title does bulk things up a bit, but yea much easier for sorting.

An alternative could be a list of unique songs by id, and a separate list mapping titles to ids.

File: app/controllers/songs_controller.rb [hunk 5]
```diff
@@ -65,48 +66,52 @@ def adjust_lang_params
...
+      lyrics: song.lyrics
+    }
   end
 
   def song_params
     params.require(:song).permit(:lyrics, :firstline_title, :custom_title, :chorus_title, :lang)
   end
 
+  def sort_songs(songs)
+    songs.sort_by! { |s| clean_for_sorting(s[:title]) }
```
[the remark this answers is not in the document]
Comment (ReganRyanNZ) [reply app/controllers/songs_controller.rb]: I have a hunch they are stripped, this might be just a sanity check. It's good for upcasing everything though (I've got some mega beef about isilo hymnal firstline index/capitalisation).

It could potentially even become ["O","Oh"] -> "O" to get rid of that nightmare
